feat(#410): add support for `LTREE` type.

landure opened the pull request:
- add `Ltree` value object
- add `Ltree` doctrine type
- add corresponding unit and integration tests

fix #410 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added PostgreSQL ltree support: DBAL type and a concrete Ltree value object with string/JSON, hierarchy helpers and immutability.

* **Bug Fixes / Errors**
  * Introduced dedicated, user-friendly exceptions for DB↔PHP conversion and value-object validation with clearer messages.

* **Documentation**
  * Added ltree usage guide, migration/indexing guidance and integration examples for Doctrine/Symfony/Laravel; updated AVAILABLE-TYPES and README.

* **Tests**
  * Added unit and integration tests and enabled ltree in the test environment.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

landure commented:
This pull request only add the `ltree` type for doctrine. I've not worked on functions.

landure commented on tests/Unit/MartinGeorgiev/Doctrine/DBAL/Types/LtreeTest.php:
```diff
@@ -0,0 +1,123 @@
+     */
+    public static function provideValidTransformations(): array
+    {
+        return [
+            'null' => [
+                'ltreeValueObject' => null,
+                'postgresValue' => null,
+            ],
+            'valid empty ltree' => [
+                'ltreeValueObject' => new LtreeValueObject([]),
+                'postgresValue' => '',
+            ],
```
I've tested this on a PostgreSQL 17.5 server. It accepts empty ltree values.

landure commented on src/MartinGeorgiev/Doctrine/DBAL/Types/ValueObject/Ltree.php:
```diff
@@ -0,0 +1,130 @@
+    #[\Override]
+    public static function fromString(string $ltree): static
+    {
+        if ('' === $ltree) {
+            return new static([]);
+        }
+
+        $branch = \explode('.', $ltree);
+
+        return new static($branch); // @phpstan-ignore-line argument.type
```
ltree accept empty string, verified in integration tests.

landure commented on src/MartinGeorgiev/Doctrine/DBAL/Types/Exceptions/InvalidLtreeForPHPException.php:
```diff
@@ -0,0 +1,20 @@
+
+class InvalidLtreeForPHPException extends ConversionException
+{
+    private static function create(string $message, mixed $value): self
+    {
+        return new self(\sprintf($message, \var_export($value, true)));
+    }
+
+    public static function forInvalidType(mixed $value): self
+    {
+        return self::create('Value must be a LtreeInterface, %s given', \gettype($value));
+    }
```
The code is a copy-paste from InvalidPointForPHPException.php . I've kept the same logic for consistancy.

landure commented on src/MartinGeorgiev/Doctrine/DBAL/Types/ValueObject/Ltree.php:
```diff
@@ -0,0 +1,136 @@
+    #[\Override]
+    public function isRoot(): bool
+    {
+        return 1 >= \count($this->pathFromRoot);
```
I've added a `isEmpty()` method to distinguish root paths from empty paths.

landure replied:
it's the case in last commit of the pull request

landure commented on docs/INTEGRATING-WITH-LARAVEL.md:
```diff
@@ -25,18 +25,18 @@
                 'real[]' => 'real[]',
                 '_float4' => 'real',
-                
+
```
this commit concerns the Ltree type, not int8 or float4

landure commented:
Hi @martin-georgiev, this is missing supports for operators, but I believe this pull request OK for minimal Ltree support. What do you think ?

martin-georgiev commented:
@coderabbitai,
Consider the edge cases in the integration and unit test coverage in this repository and the completeness of documentation about DBAL types. Is there anything outstanding you would suggest for this PR? Are all the repository styles and architecture respected?

martin-georgiev commented:
@coderabbitai - you should know that all external links in this repository are to permanent addresses like the versioned PostgreSQL documentation. Links with `/current` are fragile and may not contain relevant information in the near future. Add code suggestions based on this.

martin-georgiev commented on src/MartinGeorgiev/Doctrine/DBAL/Types/ValueObject/LtreeInterface.php:
```diff
@@ -0,0 +1,70 @@
+
+namespace MartinGeorgiev\Doctrine\DBAL\Types\ValueObject;
+
+interface LtreeInterface extends \Stringable, \JsonSerializable
```
@coderabbit, what is the reason you consider this interface naming "robust"? There is no such example in the database> Your suggestions shall be as if made by a Principal Software Engineer, not mediocre. This naming (and the interface's very need) does not seem in the spirit of the codebase.

martin-georgiev commented on src/MartinGeorgiev/Doctrine/DBAL/Types/ValueObject/Ltree.php:
```diff
@@ -0,0 +1,190 @@
+    protected static function assertValidLtreeNode(mixed $value): void
+    {
+        if (!\is_string($value) || '' === $value) {
+            throw new \InvalidArgumentException('Value must be a non-empty string.');
```
Usually, this will be a domain exception. Some example that can be used is `InvalidFloatArrayItemForDatabaseException`.

landure replied:
In `Types/Ltree.php`, this is the case.
In `Types/ValueObject`, i followed the example givenv by `Types/ValueObject/Point.php`, which throws `\InvalidArgumentException`:

```php
    public static function fromString(string $pointString): self
    {
        if (!\preg_match(self::POINT_REGEX, $pointString, $matches)) {
            throw new \InvalidArgumentException(
                \sprintf('Invalid point format. Expected format matching %s, got: %s', self::POINT_REGEX, $pointString)
            );
        }

        return new self((float) $matches[1], (float) $matches[2]);
    }
```

martin-georgiev replied:
My bad, there is an upcoming change that extracts these into new classes, but it hasn't been merged yet. I misled you, given the current state of the main branch.

martin-georgiev commented on src/MartinGeorgiev/Doctrine/DBAL/Types/ValueObject/Ltree.php:
```diff
@@ -0,0 +1,190 @@
+    #[\Override]
+    public function jsonSerialize(): array
+    {
+        return $this->pathFromRoot;
```
```suggestion
        return $this->getPathFromRoot();
```

martin-georgiev commented on src/MartinGeorgiev/Doctrine/DBAL/Types/ValueObject/Ltree.php:
```diff
@@ -0,0 +1,190 @@
+    public static function fromString(string $ltree): static
+    {
+        if ('' === $ltree) {
+            return new static([]);
```
What is the case of for late static binding? Perhaphs I miss the point.

landure replied:
I would use `self` if the class was `final`. In this case, I use `static` if someone want to `extends` the class.
What do you think ? Should I remove any `static` use ?

landure replied:
I would keep using `static`, in case someone want to add a method to `Ltree`.

martin-georgiev replied:
Ok, this sounds like a reasonable reason to resolve the convo as-is :)

martin-georgiev commented on src/MartinGeorgiev/Doctrine/DBAL/Types/ValueObject/Ltree.php:
```diff
@@ -0,0 +1,190 @@
+
+        $pathFromRoot = \explode('.', $ltree);
+
+        self::assertListOfValidLtreeNodes($pathFromRoot);
```
Why assert when this is done in the constructor?

landure replied:
the constructor expects a `list<non-empty-string>`. This assertion prevents a PHPStan alert.

There's two alternatives to this way of doing things:

- modify the constructor @param to expect `list<string>`.
- force the type of $pathFromRoot with `/** @var list<non-empty-string> $pathFromRoot */` above line 36.

which do you prefer ?

martin-georgiev replied:
Sometimes I will suppress the PHPStan error if it doesn't make sense in the code. PHPStan is great, but there are occasions where it brings more noise than value.

martin-georgiev commented on src/MartinGeorgiev/Doctrine/DBAL/Types/ValueObject/Ltree.php:
```diff
@@ -0,0 +1,190 @@
+     * @throws \InvalidArgumentException if the pathFromRoot contains empty strings or is not a list
+     */
+    public function __construct(
+        private readonly array $pathFromRoot,
```
Perhaps it be cleaner to use expanding arguments here.
```suggestion
        private readonly ...$pathFromRoot,
```

landure replied:
I would agree if the node order wasn't important. What you propose is true for something like:

```
public function addChildren(...$children) {}
```

In $pathFromRoot case, the order of the list contents has meaning: `[ 0 => 'root', 1 => 'child', 2 => 'grand-child' ]`.

martin-georgiev replied:
I am not sure I understand your concern. I'm not too bothered if this stays an array or an argument list - just saying that argument unpacking will preserve the order, and that "root -> child -> grand-child" will not somehow become "root -> grand-child -> child".

landure replied:
I would prefer it to stay an array, since it might prevent issues with Symfony serializer, which I don't know the behavior of when confronted with argument unpacking.
But I'll follow your lead on this.

martin-georgiev replied:
Ok, let's leave it as per your preference.

martin-georgiev commented on tests/Unit/MartinGeorgiev/Doctrine/DBAL/Types/LtreeTest.php:
```diff
@@ -0,0 +1,172 @@
+        self::assertSame('ltree', $this->fixture->getName());
+    }
+
+    public function test_get_sql_declaration(): void
```
The repository uses PHP attributes instead of the prefix convention for the tests. Can you please import the relevant PHPUnit `Test` class for it? Please take a look at the other test files around this one for examples.

martin-georgiev commented on src/MartinGeorgiev/Doctrine/DBAL/Types/ValueObject/Ltree.php:
```diff
@@ -0,0 +1,190 @@
+    }
+
+    #[\Override]
+    public function equals(LtreeInterface $ltree): bool
```
I will refrain from introducing this method to just one of the value objects in the repository. Sounds like it solves a problem that is not a concern of the project. If this quality check is really needed, it is better to have a consistently applied, unified style among all applicable value objects in a separate PR.

landure replied:
Removed by my last commit.

landure commented on docs/LTREE-TYPE.md:
```diff
@@ -0,0 +1,225 @@
+    private Uuid $id;
+
+    #[ORM\Column(type: 'ltree', unique: true)]
+    private Ltree $path;
```
Fixed by last push

landure commented:
I've just refactored Ltree value object to have its own InvalidArgumentException class, and fixed the integration tests to match the new method name.

martin-georgiev commented on src/MartinGeorgiev/Doctrine/DBAL/Types/Ltree.php:
```diff
@@ -0,0 +1,94 @@
+    protected const TYPE_NAME = 'ltree';
+
+    #[\Override]
+    public function getSQLDeclaration(array $column, AbstractPlatform $platform): string
+    {
+        $this->assertPostgreSQLPlatform($platform);
+
+        return parent::getSQLDeclaration($column, $platform);
+    }
+
+    #[\Override]
+    public function getMappedDatabaseTypes(AbstractPlatform $platform): array
+    {
+        $this->assertPostgreSQLPlatform($platform);
+
+        return [
+            'ltree',
+        ];
+    }
```
No other types in the repository require this. Why is it needed?

martin-georgiev commented on src/MartinGeorgiev/Doctrine/DBAL/Types/Ltree.php:
```diff
@@ -0,0 +1,94 @@
+    /**
+     * {@inheritDoc}
+     *
+     * @SuppressWarnings("PHPMD.StaticAccess")
```
Where does this come from? PHP Mess Detector is not a requirement of this project.

landure replied:
I wrote this class in my dev environment. fixing.

martin-georgiev commented on tests/Integration/MartinGeorgiev/Doctrine/DBAL/Types/LtreeTypeTest.php:
```diff
@@ -0,0 +1,99 @@
+        $this->assertSame((string) $ltreeValueObject, (string) $actual, 'Failed asserting that ltree string representations are identical for type '.$typeName);
+    }
+
+    private function createLtreeExtension(): void
+    {
+        $sql = 'CREATE EXTENSION IF NOT EXISTS ltree';
+        $this->connection->executeStatement($sql);
+    }
```
We already centralize extension setup in `tests/Integration/MartinGeorgiev/TestCase.php`. Please hook this into the [createTestSchema()](https://github.com/martin-georgiev/postgresql-for-doctrine/blob/d97caae042ecbdd1e958b512e38ca4d56fd3d711/tests/Integration/MartinGeorgiev/TestCase.php#L168) mechanics instead of duplicating it here. This keeps initialization consistent and reduces maintenance.

martin-georgiev commented:
@coderabbit, pay extra attention to the new tests setup and the exception logic. Anything that you notice that is off from the codebase styles and patterns?

landure commented on src/MartinGeorgiev/Doctrine/DBAL/Types/ValueObject/Ltree.php:
```diff
@@ -0,0 +1,201 @@
+        return new static($parentPathFromRoot);
+    }
+
+    /**
+     * Checks if the ltree has no nodes.
+     */
```
This one comment was placed to appease coderabbit. It's aims is to clearly differentiate a root path, with one node, from an empty path, with no node. I'll remove them if you prefer.

landure replied:
removed this docblock and another one that had a redondant comment.

landure commented on src/MartinGeorgiev/Doctrine/DBAL/Types/Ltree.php:
```diff
@@ -0,0 +1,94 @@
+            return;
+        }
+
+        throw new \LogicException('Ltree DBAL type can only be used with the PostgreSQL platform.');
```
I wrote this class outside the project before importing it. Honestly, this is overengineering on my part.

landure replied:
Removed.

landure commented on src/MartinGeorgiev/Doctrine/DBAL/Types/ValueObject/Exceptions/InvalidLtreeException.php:
```diff
@@ -0,0 +1,37 @@
+    public static function forInvalidPathFromRootFormat(array $value, string $expectedFormat): self
+    {
+        return new self(\sprintf(
+            "Invalid Ltree's path from root format. Expected %s, got: %s",
```
Done

martin-georgiev commented on src/MartinGeorgiev/Doctrine/DBAL/Types/ValueObject/Ltree.php:
```diff
@@ -0,0 +1,201 @@
+            throw InvalidLtreeException::forInvalidNodeFormat($value, 'non-empty string');
+        }
+
+        if (\str_contains($value, '.')) {
```
How silly of me! I misread the code and missed that here $value is wiyh the emaning of a leaf, and not the meaningn of the entire tree structure.

martin-georgiev commented on src/MartinGeorgiev/Doctrine/DBAL/Types/ValueObject/Ltree.php:
```diff
@@ -0,0 +1,196 @@
+    public function getParent(): static
+    {
+        if ($this->isEmpty()) {
+            throw new \LogicException('Empty ltree has no parent.');
```
This shall use `InvalidLtreeException`.

landure replied:
Hi, I made the change, but `InvalidLtreeException` extends `InvalidArgumentException`.  Here it's either a `LogicException` since the code should never get the parent of an empty `Ltree`, or `RuntimeException`, since this only occurs at runtime.

martin-georgiev commented on src/MartinGeorgiev/Doctrine/DBAL/Types/Ltree.php:
```diff
@@ -0,0 +1,55 @@
+        if (\is_string($value)) {
+            try {
+                return LtreeValueObject::fromString($value);
+            } catch (\InvalidArgumentException) {
```
This shall catch `InvalidLtreeException` only.

martin-georgiev commented on src/MartinGeorgiev/Doctrine/DBAL/Types/Ltree.php:
```diff
@@ -0,0 +1,55 @@
+        if (\is_string($value)) {
+            try {
+                $value = LtreeValueObject::fromString($value);
+            } catch (\InvalidArgumentException) {
```
This shall catch `InvalidLtreeException` only.

martin-georgiev commented on src/MartinGeorgiev/Doctrine/DBAL/Types/Ltree.php:
```diff
@@ -0,0 +1,55 @@
+use MartinGeorgiev\Doctrine\DBAL\Types\Exceptions\InvalidLtreeForPHPException;
+use MartinGeorgiev\Doctrine\DBAL\Types\ValueObject\Ltree as LtreeValueObject;
+
+final class Ltree extends BaseType
```
```suggestion
/**
 * Implementation of PostgreSQL LTREE data type.
 *
 * @see https://www.postgresql.org/docs/13/ltree.html
 * @since 3.5
 *
 * @author Pierre-Yves Landuré
 */
final class Ltree extends BaseType
```

martin-georgiev commented on src/MartinGeorgiev/Doctrine/DBAL/Types/ValueObject/Exceptions/InvalidLtreeException.php:
```diff
@@ -0,0 +1,37 @@
+ * This exception is specifically for validation errors within the Ltree value object itself,
+ * separate from DBAL conversion exceptions.
+ *
+ * @since 3.5
```
```suggestion
 * @since 3.5
 *
 * @author Pierre-Yves Landuré
```

landure commented:
Sorry for thé delay. I'll do thé work tommorow

landure commented:
@martin-georgiev This should be OK. I'm available until sunday to fix issues.

martin-georgiev commented on tests/Unit/MartinGeorgiev/Doctrine/DBAL/Types/ValueObject/LtreeTest.php:
```diff
@@ -0,0 +1,526 @@
+    {
+        $childAsString = (string) $child;
+        $ltree = $child->getParent();
+        self::assertNotSame($child, $ltree, 'getParent() should return a new instance');
```
Please use the instance instead of calling the assertion methods statically. All tests have teh style of `$this->assert...` instead of `self::assert...`.

landure replied:
can't believe I've forgotten to fix this.

martin-georgiev commented on tests/Unit/MartinGeorgiev/Doctrine/DBAL/Types/ValueObject/LtreeTest.php:
```diff
@@ -0,0 +1,526 @@
+
+    #[DataProvider('provideParentRelationship')]
+    #[Test]
+    public function respect_immutability_when_getting_parent(Ltree $child, Ltree $parent): void
```
I don't understand the purpose of this test or its intended assertions intentions. The parent is not used at all, but it is passed as an argument. Can you please elaborate on them?

martin-georgiev commented on tests/Unit/MartinGeorgiev/Doctrine/DBAL/Types/ValueObject/LtreeTest.php:
```diff
@@ -0,0 +1,526 @@
+    }
+
+    #[Test]
+    public function throws_exception_when_getting_empty_ltree_parent(): void
```
```suggestion
    public function throws_exception_when_ltree_is_a_root_and_therefore_has_no_parent(): void
```

landure replied:
I differentiate root tree (with one node, as given in ltree documentation example) from empty ltree (with no node), which should be the exception in the real world, even if it's allowed by the ltree extension.

Would `throws_exception_when_ltree_is_empty_and_therefore_has_no_parent` be ok ?

landure replied:
FIY, the parent of a root ltree is an empty ltree.

martin-georgiev commented on src/MartinGeorgiev/Doctrine/DBAL/Types/ValueObject/Ltree.php:
```diff
@@ -0,0 +1,205 @@
+        return new static($parentPathFromRoot);
+    }
+
+    public function isEmpty(): bool
```
What's the purpose of having support for empty instances? It is not a logical tree in this case, is it?
